### PR TITLE
Uhci driver

### PR DESCRIPTION
**Pull request by FilippChubukov**
implementation of uhci host controller driver, correctly works with usb flash device

### CODE REVIEW

File: src/drivers/usb/hc/uhci_pci.c [hunk 1]
```diff
@@ -0,0 +1,568 @@
...
+		}
+		break;
+
+    case USB_CLEAR_PORT_FEATURE:
+  		switch (ctrl->w_value) {
+  		case USB_PORT_FEATURE_POWER:
+  			break;
+  		case USB_PORT_FEATURE_C_CONNECTION:
+
+        if (ctrl->w_index == 1) {
```
Comment (anton-bondarev) [inline src/drivers/usb/hc/uhci_pci.c:218]: format please

File: src/drivers/usb/hc/uhci_pci.c [hunk 2]
```diff
@@ -0,0 +1,568 @@
...
+  		case USB_PORT_FEATURE_C_RESET:
+
+  			break;
+  		default:
+  			log_error("Unknown port clear feature: 0x%x\n", ctrl->w_value);
+  			return -1;
+  		}
+		break;
+	default:
+    	panic("uhci_root_hub_control: Unknown req_type=0x%x, request=0x%x\n",
```
Comment (anton-bondarev) [inline src/drivers/usb/hc/uhci_pci.c:237]: correct indents please

File: src/drivers/usb/hc/uhci_pci.c [hunk 3]
```diff
@@ -0,0 +1,568 @@
...
+	struct uhci_link *p = a->prev;
+	struct uhci_link *n = a;
+	n->prev = x;
+	x->next = n;
+	x->prev = p;
+	p->next = x;
+}
+
+static void uhci_remove_qh(struct uhci_qh *qh) {
+	struct uhci_qh *prev = (struct uhci_qh *)((char *)(qh->qh_link.prev) - (unsigned long)(&(((struct uhci_qh*)0)->qh_link))); /* find last active qh, TODO: rewrite this place with dlist from embox */
```
Comment (anton-bondarev) [inline src/drivers/usb/hc/uhci_pci.c:305]: format please

**Comment by anton-bondarev** [comment]
Thanks!
Has the PR description been still valid?
I mean this one
> i added printk in func uhci_process_qh, we see that the td that is second in the queue is not executed, the host controller marks it as invalid and stops
> if you check the first td in the queue, it succeeds and its reference to the next object in the queue is valid 

Also, I tried to execute as you described and got the following error:
```
Embox kernel start
	unit: initializing embox.kernel.task.task_resource: done
	unit: initializing embox.kernel.task.task_table: done
	unit: initializing embox.kernel.task.kernel_task: done
	unit: initializing embox.driver.periph_memory_mmap: done
	unit: initializing embox.arch.x86.mmu: done
	unit: initializing embox.mem.phymem: 
			start=0x0a54f000 end=0x10100000 size=96145408
		done
runlevel: init level is 0
	unit: initializing embox.mem.static_heap: done
	unit: initializing embox.kernel.sched.sched: done
	unit: initializing embox.profiler.tracing: done
	unit: initializing embox.fs.rootfs: done
	unit: initializing embox.driver.flash.flash_nofs: done
	unit: initializing embox.fs.buffer_cache: done
	unit: initializing embox.driver.interrupt.i8259: done
	unit: initializing embox.driver.clock.pit: done
	unit: initializing embox.kernel.time.kernel_time: done
	unit: initializing embox.kernel.time.timekeeper: done
	unit: initializing embox.driver.usb.core: done
	unit: initializing embox.driver.input.core: done
	unit: initializing embox.driver.usb.class.hid: done
	unit: initializing embox.driver.usb.class.ccid: done
	unit: initializing embox.driver.usb.class.mass_storage: done
	unit: initializing embox.driver.ide: done
	unit: initializing embox.driver.net.usbnet: done
	test: running embox.test.math.math_test .. done
runlevel: init level is 1
	unit: initializing embox.net.neighbour: done
	unit: initializing embox.mem.slab: done
	unit: initializing embox.driver.usb.hub: done
	unit: initializing embox.driver.pci: done
	unit: initializing embox.net.tcp: done
	unit: initializing embox.driver.tty.serial: done
	pci: Intel UHCI usb host driver inserted
EXCEPTION [0xe]: error = 00000002
EAX=fffffff4    EBX=00000000 ECX=09b86000 EDX=09b86022
 GS=00000010     FS=00000010  ES=00000010  DS=00000010
EDI=00000012    ESI=00000012 EBP=01b82ab8 EIP=00108150
 CS=00000008 EFLAGS=00010206 ESP=01b82ac8  SS=00107a62


 --   00001000 * A R W  thread 0  task 1 --------------------------------------

 25 0x0014d898 <exception_handler+0x113> ...h/x86/kernel/exception_handler.c:20
 24 0x00100079 <excep_stub+0x1c>      /src/arch/x86/kernel/exception_entry.S:49
 23 0x00108150 <uhci_insert_qh+0x2f>         /src/drivers/usb/hc/uhci_pci.c:285
 22 0x0010872f <uhci_request_do+0x25>        /src/drivers/usb/hc/uhci_pci.c:504
 21 0x00108789 <uhci_request+0x11>           /src/drivers/usb/hc/uhci_pci.c:523
 20 0x00155fd0 <usb_endp_request+0x52>          /src/drivers/usb/usb_core.c:101
 19 0x00156241 <usb_endp_control_wait+0xa2>     /src/drivers/usb/usb_core.c:171
 18 0x0015c1df <usb_device_init+0x31>            /src/drivers/usb/usb_hub.c:223
 17 0x0015c464 <usb_hub_port_init+0xa1>          /src/drivers/usb/usb_hub.c:265
 16 0x0015bc74 <usb_new_device+0x12e>             /src/drivers/usb/usb_hub.c:49
 15 0x0015c54e <usb_hub_port_event+0xa4>         /src/drivers/usb/usb_hub.c:287
 14 0x0015c58f <usb_hub_event+0x30>              /src/drivers/usb/usb_hub.c:304
 13 0x0015ca4e <usb_hub_probe+0x363>             /src/drivers/usb/usb_hub.c:352
 12 0x00157657 <usb_driver_find+0x70>          /src/drivers/usb/usb_driver.c:98
 11 0x001576fb <usb_driver_probe+0x11>        /src/drivers/usb/usb_driver.c:123
 10 0x0015bda4 <usb_new_device+0x25e>             /src/drivers/usb/usb_hub.c:49
  9 0x00107d9e <uhci_start+0x94>             /src/drivers/usb/hc/uhci_pci.c:136
  8 0x001566a1 <usb_hcd_register+0x10c>         /src/drivers/usb/usb_core.c:298
  7 0x00108855 <uhci_pci_init+0x59>          /src/drivers/usb/hc/uhci_pci.c:554
  6 0x00155237 <pci_drv_probe+0x57>            /src/drivers/pci/pci_driver.c:21
  5 0x001552c8 <pci_mod_enable+0x6c>           /src/drivers/pci/pci_driver.c:35
  4 0x0014d4e6 <mod_enable+0x7f>                  /src/framework/mod/core.c:130
  3 0x0014cfb2 <runlevel_set+0xa4>                 /src/framework/runlevel.c:54
  2 0x0010197f <init+0x28>                                /src/kernel/init.c:53
  1 0x00101933 <kernel_start+0x10>                        /src/kernel/init.c:23

run 0x00150d2b <boot_stub+0x0>               /src/kernel/sched/boot_thread.c:20

```

**Comment by FilippChubukov** [comment]
changed PR description, fixed this error

**Comment by anton-bondarev** [comment]
> changed PR description, fixed this error

Thanks. It works for me.

Please remove changes in x86/qemu template and update [USB wiki page ](https://github.com/embox/embox/wiki/Using-USB-on-qemu)

**Review by anton-bondarev — APPROVED** [review]
LGTM
I've checked it with ./scripts/qemu/auto_qemu_with_usb_storage script

**Comment by anton-bondarev** [comment]
@FilippChubukov . Thanks. I've merged it

**Pull request merged**